### PR TITLE
definition templates v0

### PR DESCRIPTION
**Pull request by psunnn**
# Definition Templates v0

## Database
- Add `definition_template` table with the following columns:
  - `id VARCHAR PRIMARY KEY`
  - `type VARCHAR`
  - `version INTEGER`
  - `schema JSONB`: the JSON schema
  - `template TEXT`: go template string
  - `image VARCHAR`: docker image
- Add the following columns to the `task_def` and `task` tables:
  - `template_id VARCHAR REFERENCES task_type(id)`
  - `template_payload JSONB`

## Endpoints
New V7 endpoints:
- `GET /api/v7/template`: list all templates
- `GET /api/v7/template?latest_only=true`: list latest version for each template type
- `POST /api/v7/template`: create new template
- `GET /api/v7/template/:template_id`: describe template

## Flow
During the task creation process, users will be able to select a `task_type`. The UI will then render a custom sub-form based on the JSON schema of the selected task type. The results of this sub-form will be stored as the definition's `payload` attribute.

During task execution, the execution service will take the definition's payload, validate it against the task type's JSON schema (via a lib like https://github.com/xeipuuv/gojsonschema; the validation step can also be performed at task creation time), and dump the payload into the task type's template string, which will be used as the run's `command` property.

For v0, task type schema/template creation will be managed by WEE. In subsequent iterations, admins will be able to create custom task types by writing JSON schemas and template strings.

Note: this will not affect legacy task definitions.

### CODE REVIEW

File: state/pg_queries.go [hunk 1]
```diff
@@ -309,3 +321,10 @@ const GetWorkerSQL = WorkerSelect + "\nwhere worker_type = $1 and engine = $2"
 // worker type; locks the row.
 //
 const GetWorkerSQLForUpdate = GetWorkerSQL + " for update"
+
+const taskTypeSelect = `
+  SELECT id, alias, schema::TEXT, template
+  FROM task_type
+`
+const ListTaskTypeSQL = taskTypeSelect + "\n limit $1 offset $2"
```
Comment (ujjwalsfix) [inline state/pg_queries.go]: Orthogonally ties into versioning, how would multiple versions of the same task? get latest or last version?

File: testutils/mocks.go [hunk 2]
```diff
@@ -389,3 +390,15 @@ func (iatt *ImplementsAllTheThings) Logs(definition state.Definition, run state.
 	iatt.Calls = append(iatt.Calls, "Logs")
 	return "", nil, nil
 }
+
+// ListWorkers - StateManager
+func (iatt *ImplementsAllTheThings) ListTaskTypes(limit int, offset int) (state.TaskTypeList, error) {
+	iatt.Calls = append(iatt.Calls, "ListTaskTypes")
+	return state.TaskTypeList{Total: len(iatt.TaskTypes), TaskTypes: iatt.TaskTypes}, nil
```
Comment (ujjwalsfix) [inline testutils/mocks.go]: Nice - `len()`

File: services/definition.go [hunk 3]
```diff
@@ -49,8 +50,28 @@ func NewDefinitionService(conf config.Config, ecsExecutionEngine engine.Engine,
...
+
+		if err != nil {
+			return state.Definition{}, err
+		}
+
+		schemaLoader := gojsonschema.NewStringLoader(taskType.Schema)
+		documentLoader := gojsonschema.NewStringLoader(fmt.Sprintf("%v", definition.TemplatePayload))
+		result, err := gojsonschema.Validate(schemaLoader, documentLoader)
+
+		fmt.Println(result)
```
Comment (psunnn) [inline services/definition.go]: todo: remove

**Comment by psunnn** [comment]
closing this PR in favor of #283

**Pull request closed without merging**